### PR TITLE
WIP assign unique id stream properties

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java [hunk 1]
```diff
@@ -546,7 +558,7 @@ public StreamProperties visitSample(SampleNode node, List<StreamProperties> inpu
 
         private final Optional<List<Symbol>> partitioningColumns; // if missing => partitioned with some unknown scheme
 
-        private final boolean ordered;
+        private final Optional<OrderingScheme> ordering;
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java]: Right. Makes sense

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java [hunk 2]
```diff
@@ -433,7 +435,15 @@ public StreamProperties visitEnforceSingleRow(EnforceSingleRowNode node, List<St
         @Override
         public StreamProperties visitAssignUniqueId(AssignUniqueId node, List<StreamProperties> inputProperties)
         {
-            return Iterables.getOnlyElement(inputProperties);
+            StreamProperties sourceProperties = Iterables.getOnlyElement(inputProperties);
+            return new StreamProperties(
+                    sourceProperties.getDistribution(),
+                    Optional.of(node.getOutputSymbols()),
+                    Optional.of(new OrderingScheme(
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java]: @sopel39 @martint 

`OrderingScheme` class contains two fields: `orderBy` and `orderings`. What's the difference between them? Why do we need `orderBy`? Isn't `orderBy` same as `orderings.keys()`?

```
public class OrderingScheme
{
    private final List<Symbol> orderBy;
    private final Map<Symbol, SortOrder> orderings;
```

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java]: @mbasmanova 
`orderBy` list preservers order of sort columns

`orderings` map are created via:
```
this.orderings = ImmutableMap.copyOf(orderings);
```
which means `keys()` will be returned in the same order as entries from copied `orderings` map. We shouldn't assume that map that we copy from will return keys in the same order as `orderBy` list

Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java]: @sopel39 Karol, thanks for explaining. It seems more natural to me to use parallel lists: `List<Symbol>` and `List<SortOrder>` or a list of pairs, but I understand why there are two now.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java]: @martint @sopel39 Are we going to delete `boolean ordered;` flag and update the code to set and check `otherActualProperties.local.grouping(cols)` instead?

[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java]: @mbasmanova I think we could delete `ordered` flag, but we need to validate if it has exactly the same semantics as `otherActualProperties.local.grouping(cols)` first.

**Comment by sopel39** [comment]
@mbasmanova I'm closing this one as maybe we want to use `otherActualProperties.local.grouping(cols)` instead of `ordered` flag as you mentioned. Feel free to pick it up from here. The concept for `AssignUniqueId` stream properties support should be similar as to the one here

**Pull request closed without merging**